README
= validates_timeliness

  * Source:  http://github.com/adzap/validates_timeliness
  * Bugs:    http://adzap.lighthouseapp.com/projects/14111-validates_timeliness

== DESCRIPTION:

Validate dates, times and datetimes for Rails 2.x. Plays nicely with new Rails 2.1
features such as automatic timezone handling and dirty attributes. Allows you to
add custom formats or remove defaults easily. This allows you to control what you 
think should be a valid date or time string.


== FEATURES:

* Adds ActiveRecord validation for dates, times and datetimes

* Add or remove date/time formats to customize validation

* Create new formats using very simple date/time format tokens

* Restores ability to see raw value entered for date/time attributes with
  _before_type_cast modifier, which was lost in Rails 2.1.

* Respects new timezone features of Rails 2.1.


== INSTALLATION:

  ./script/plugin git://github.com/adzap/validates_timeliness

  
== USAGE:

To validate a model with a date, time or datetime attribute you just use the 
validation method

  class Person < ActiveRecord::Base
    validates_date :date_of_birth

  end
  
The list of validation methods available are as follows:

  * validates_date     - validate value as date
  
  * validates_time     - validate value as time only i.e. '12:20pm'
  
  * validates_datetime - validate value as a full date and time
  
The validation methods take the usual options plus some specific ones to restrict
the valid range of dates or times allowed

  Temporal options (or restrictions):  
    :before       - Attribute must be before this value to be valid
    :on_or_before - Attribute must be equal to or before this value to be valid
    :after        - Attribute must be after this value to be valid
    :on_or_after  - Attribute must be equal to or after this value to be valid

  Regular validation options:  
    :allow_nil    - Allow a nil value to be valid
    :allow_blank  - Allows a nil or empty string value to be valid
    :if           - Execute validation when :if evaluates true
    :unless       - Execute validation when :unless evaluates false

  Message options: - Use these to override the default error messages
    :invalid_date_message
    :invalid_time_message
    :invalid_datetime_message
    :before_message
    :on_or_before_message
    :after_message
    :on_or_after_message

The temporal restrictions can take 4 different value types:

  * String value
  * Date, Time, or DateTime object value
  * Proc or lambda object
  * A symbol matching the method name in the model  

When an attribute value is compared to temporal restrictions, they are compared as 
the same type as the validation method type. So using validates_date means all
values are compared as dates.

== EXAMPLES:

  validates_date :date_of_birth :before => Proc.new { 18.years.ago },
                                :before_message => "must be at least 18 years old"
  
  validates_time :breakfast_time, :on_or_after => '6:00am',
                                  :on_or_after_message => 'must be after opening time',
                                  :before => :second_breakfast_time,
                                  :allow_nil => true

  validates_datetime :appointment_date, :before => Proc.new { 1.week.from_now }
 

=== DATE/TIME FORMATS:

So what formats does the plugin allow. Well there are default formats which can 
be added to easily using the plugins format rules. Also formats can be easily 
removed without hacking the plugin at all. 

Below are the default formats. If you think they are easy to read then you will 
be happy to know that is exactly the format you can use to define your own if 
you want. No complex regular expressions or duck punching (monkey patching) the
plugin is needed.

  Time formats:  
    hh:nn:ss
    hh-nn-ss
    h:nn
    h.nn
    h nn
    h-nn
    h:nn_ampm
    h.nn_ampm 
    h nn_ampm
    h-nn_ampm
    h_ampm
  
  NOTE: Any time format without a meridian token (the 'ampm' token) is considered 
        in 24 hour time.
  
  Date formats:  
    yyyy/mm/dd
    yyyy-mm-dd
    yyyy.mm.dd
    m/d/yy  OR  d/m/yy
    m\d\yy  OR  d\m\yy
    d-m-yy
    d.m.yy
    d mmm yy
  
  NOTE: To use non-US date formats see US/EURO FORMATS section
  
  Datetime formats:  
    m/d/yy h:nn:ss   OR  d/m/yy hh:nn:ss
    m/d/yy h:nn      OR  d/m/yy h:nn
    m/d/yy h:nn_ampm OR  d/m/yy h:nn_ampm
    yyyy-mm-dd hh:nn:ss
    yyyy-mm-dd h:nn
    ddd mmm d hh:nn:ss zo yyyy # Ruby time string
    yyyy-mm-ddThh:nn:ss(?:Z|zo) # ISO 8601

  NOTE: To use non-US date formats see US/EURO FORMATS section

Here is what each format token means:

  Format tokens:   
       y = year
       m = month
       d = day
       h = hour
       n = minute
       s = second
       u = micro-seconds
    ampm = meridian (am or pm) with or without dots (e.g. am, a.m, or a.m.)
       _ = optional space
      tz = Timezone abbreviation (e.g. UTC, GMT, PST, EST)
      zo = Timezone offset (e.g. +10:00, -08:00, +1000)

  Repeating tokens:        
       x = 1 or 2 digits for unit (e.g. 'h' means an hour can be '9' or '09')
      xx = 2 digits exactly for unit (e.g. 'hh' means an hour can only be '09')
      
  Special Cases:
      yy = 2 or 4 digit year
   yyyyy = exactly 4 digit year
     mmm = month long name (e.g. 'Jul' or 'July')
     ddd = Day name of 3 to 9 letters (e.g. Wed or Wednesday)
       u = microseconds matches 1 to 3 digits

All other characters are considered literal. For the technically minded 
(well you are developers), these formats are compiled into regular expressions 
at runtime so don't add any extra overhead than using regular expressions 
directly. So, no, it won't make your app slow!

To see all defined formats look in lib/validates_timeliness/formats.rb.

=== US/EURO FORMATS

The perenial problem for non-US developers or applications not primarily for the
US, is the US date format of m/d/yy. This is ambiguous with the European format
of d/my/yy. By default the plugin uses the US formats as this is the Ruby default
when it does date interpretation, and is in keeping PoLS (principle of least
surprise). 

To switch to using the European (or Rest of The World) formats put this in an
initializer or environment.rb

  ValidatesTimeliness::Formats.remove_us_formats  

Now '01/02/2000' will be parsed as 1st February 2000, instead of 2nd January 2000.

=== CUSTOMISING FORMATS:

I hear you say "Thats greats but I don't want X format to be valid". Well to 
remove a format stick this in an initializer file

  ValidatesTimeliness::Formats.remove_formats(:date, 'm\d\yy')

Done! That format is no longer considered valid. Easy!

Ok, now I hear you say "Well I have format that I want to use but you don't have it".
Ahh, then add it yourself. Again stick this in an initializer file

    ValidatesTimeliness::Formats.add_formats(:time, "d o'clock")

Now "10 o'clock" will be a valid value. So easy, no more whingeing!

You can embed regular expressions in the format but no gurantees that it will 
remain intact. If you avoid the use of any token characters and regexp dots or 
backslashes as special characters in the regexp, it may well work as expected. 
For special characters use POSIX character classes for safety. See the ISO 8601 
datetime for an example of an embedded regular expression.

Because formats are evaluated in order, adding a format which may be ambiguous 
with an existing format, will mean your format is ignored. If you need to make
your new format higher precedence than an existing format, you can include the 
before option like so

    ValidatesTimeliness::Formats.add_formats(:time, 'ss:nn:hh', :before => 'hh:nn:ss')

Now a time of '59:30:23' will be interpreted as 11:30:59 pm. This option saves
you adding a new one and deleting an old one to get it to work.


=== TEMPORAL RESTRICTION ERRORS:

When using the validation temporal restrictions there are times when the restriction
value itself may be invalid. Normally this will add an error to the model such as
'restriction :before value was invalid'. These can be annoying if you are using 
procs or methods as restrictions and don't care if they don't evaluate properly
and you want the validation to complete. In these situations you turn them off.

To turn them off:

  ValidatesTimeliness.ignore_restriction_errors = true

A word of warning though, as this may hide issues with the model and make those
corner cases a little harder to test. In general if you are using procs or 
model methods and you only care when they return  a value, then they should 
return nil in all other situations. Restrictions are skipped if they are nil.

=== OTHER CUSTOMISATION:

The error messages for each temporal restrictions can also be globally overridden by 
updating the default AR error messages like so

For Rails 2.0/2.1:

  ActiveRecord::Errors.default_error_messages.update(
    :invalid_date     => "is not a valid date",
    :invalid_time     => "is not a valid time",
    :invalid_datetime => "is not a valid datetime",
    :before           => "must be before %s",
    :on_or_before     => "must be on or before %s",
    :after            => "must be after %s",
    :on_or_after      => "must be on or after %s"
  )

Where %s is the interpolation value for the restriction.

Rails 2.2+ using the I18n system to define new defaults:

  en:
    activerecord:
      errors:
        messages:
          on_or_before: "must equal to or before {{restriction}}"
          on_or_after:  "must equal to or after {{restriction}}"

The {{restriction}} signifies where the interpolation value for the restriction
will be inserted.

And for something a little more specific you can override the format of the interpolation
values inserted in the error messages for temporal restrictions like so

  ValidatesTimeliness.error_value_formats.update(
    :time     => '%H:%M:%S',
    :date     => '%Y-%m-%d',
    :datetime => '%Y-%m-%d %H:%M:%S'
  )

Those are Ruby strftime formats not the plugin formats.


=== RSPEC MATCHER:

To sweeten the deal that little bit more, you have an Rspec matcher available for
you model specs. Now you can easily test the validations you have just written 
with the plugin or better yet *before* you write them! You just use the 
validation options you want as you would with the validation method. Those 
options are then verified and reported if they fail. Use it like so:

   @person.should validate_date(:birth_date, :before => Time.now, :before_message => 'should be before today')


The matcher names are just the singular of the validation methods.

== CREDITS:

* Adam Meehan (adam.meehan@gmail.com, http://duckpunching.com/)

* Jonathan Viney (http://workingwithrails.com/person/4985-jonathan-viney)
  For his validates_date_time plugin which I have used up until this plugin and
  which influenced some of the design and I borrowed a little of code from it.


== LICENSE:

Copyright (c) 2008 Adam Meehan, released under the MIT license


CHANGELOG
= 2.3.2 [2010-11-07]
- Fixed parser for string with timezone offset (thanks sigi)
- Support for latest I18n interpolation tokens in locale file
- Fixed parser allowing an hour over 12 for AM meridian

= 2.3.1 [2010-03-19]
- Fixed bug where custom attribute writer method for date/times were being overriden

= 2.3.0 [2010-02-04]
- Backwards incompatible change to :equal_to option. Fixed error message clash with :equal_to option which exists in Rails already. Option is now :is_at.
- Fixed I18n support so it returns missing translation message instead of error
- Fixed attribute method bug. Write method was bypassed when method was first generated and used Rails default parser.
- Fixed date/time selects when using enable_datetime_select_extension! when some values empty
- Fixed ISO8601 datetime format which is now split into two formats
- Changed I18n error value format to fallback to global default if missing in locale
- Refactored date/time select invalid value extension to use param values. Functionality will be extracted from plugin for v3.

= 2.2.2 [2009-09-19]
- Fixed dummy_time using make_time to respect timezone. Fixes 1.9.1 bug.

= 2.2.1 [2009-09-12]
- Fixed dummy date part for time types in Validator.type_cast_value
- No more core extensions! Removed dummy_time methods.

= 2.2.0 [2009-09-12]
- Ruby 1.9 support!
- Customise dummy date values for time types. See DUMMY DATE FOR TIME TYPES.
- Fixed matcher conflict with Shoulda. Load plugin matcher manually now see matcher section in README
- Fixed :ignore_usec when used with :with_time or :with_date
- Some clean up and refactoring

= 2.1.0 [2009-06-20]
- Added ambiguous year threshold setting in Formats class to customize the threshold for 2 digit years (See README)
- Fixed interpolation values in custom error message for Rails 2.2+
- Fixed custom I18n local override of en locale
- Dramatically simplified ActiveRecord monkey patching and hackery

= 2.0.0 [2009-04-12]
- Error value formats are now specified in the i18n locale file instead of updating plugin hash. See OTHER CUSTOMISATION section in README.
- Date/time select helper extension is disabled by default. To enable see DISPLAY INVALID VALUES IN DATE HELPERS section in README to enable.
- Added :format option to limit validation to a single format if desired
- Matcher now supports :equal_to option
- Formats.parse can take :include_offset option to include offset value from string in seconds, if string contains an offset. Offset not used in rest of plugin yet.
- Refactored to remove as much plugin code from ActiveRecord as possible.

= 1.1.7 [2009-03-26]
- Minor change to multiparameter attributes which I had not properly implemented for chaining

= 1.1.6 [2009-03-19]
- Rail 2.3 support
- Added :with_date and :with_time options. They allow an attribute to be combined with another attribute or value to make a datetime value for validation against the temporal restrictions
- Added :equal_to option
- Option key validation
- Better behaviour with other plugins using alias_method_chain on read_attribute and define_attribute_methods
- Added option to enable datetime_select extension for future use to optionally enable. Enabled by default until version 2.
- Added :ignore_usec option for datetime restrictions to be compared without microsecond
- some refactoring

= 1.1.5 [2009-01-21]
- Fixed regex for 'yy' format token which wasn't greedy enough for date formats ending with year when a datetime string parsed as date with a 4 digit year

= 1.1.4 [2009-01-13]
- Make months names respect i18n in Formats

= 1.1.3 [2009-01-13]
- Fixed bug where time and date attributes still being parsed on read using Rails default parser [reported by Brad (pvjq)]

= 1.1.2 [2009-01-12]
- Fixed bugs
 - matcher failing for custom error message without interpolation keys using I18n
 - validator custom error messages not being extracted

= 1.1.1 [2009-01-03]
- Fixed bug in matcher for options local variable

= 1.1.0 [2009-01-01]
- Added between option

= 1.0.0 [2008-12-06]
- Gemified!
- Refactor of plugin into a Data Mapper style validator class which makes for a cleaner implementation and possible future Merb\Data Mapper support
- Added Rails 2.2 i18n support. Plugin error messages can specified in locale files. See README.
- ignore_datetime_restriction_errors setting has been moved from AR to ValidatesTimeliness::Validator.ignore_restriction_errors
- date_time_error_value_formats setting has been moved from AR to ValidatesTimeliness::Validator.error_value_formats
- Namespaced modules and specs
- Clean up of specs
- fixed a few bugs
  - accessor methods not generating properly due method name stored as symbol in generated_attributes which fails on lookup
  - force value assigned to time/datetime attributes to time objects

= 0.1.0 [2008-12-06]
- Tagged plugin as version 0.1.0

= 2008-11-13
- allow uppercase meridian to be valid [reported by Alex (http://alex.digns.com/)]

= 2008-10-28
- fixed bug when dirty attributes not reflecting change when attribute changed from time value to nil [reported by Brad (pvjq)]
- fixes for Rails 2.2 compatibility. Will refactor in to Rails version specific branches in the future.

= 2008-09-24
- refactored attribute write method definitions

= 2008-08-25
- fixed bug for non-timezone write method not updating changed attributes hash [reported by Sylvestre Mergulhão]

= 2008-08-22
- fixed bug with attribute cache not clearing on write for date and time columns [reported by Sylvestre Mergulhão]
- parse method returns Date object for date column assigned string as per normal Rails behaviour
- parse method returns same object type when assigned Date or Time object as per normal Rails behaviour

= 2008-08-07
- modified matcher option value parsing to allow same value types as validation method
- fixed matcher message

= 2008-08-02
- refactored validation
- refactored matcher

= 2008-07-30
- removed setting values to nil when validation fails to preserve before_type_cast value


lib/validates_timeliness/formats.rb
require 'date'

module ValidatesTimeliness

  # A date and time parsing library which allows you to add custom formats using
  # simple predefined tokens. This makes it much easier to catalogue and customize
  # the formats rather than dealing directly with regular expressions.
  #
  # Formats can be added or removed to customize the set of valid date or time
  # string values.
  #
  class Formats
    cattr_accessor :time_formats,
                   :date_formats,
                   :datetime_formats,
                   :time_expressions,
                   :date_expressions,
                   :datetime_expressions,
                   :format_tokens,
                   :format_proc_args


    # Set the threshold value for a two digit year to be considered last century
    #
    # Default: 30
    #
    #   Example:
    #     year = '29' is considered 2029
    #     year = '30' is considered 1930
    #
    cattr_accessor :ambiguous_year_threshold
    self.ambiguous_year_threshold = 30

    # Set the dummy date part for a time type value. Should be an array of 3 values
    # being year, month and day in that order.
    #
    # Default: [ 2000, 1, 1 ] same as ActiveRecord
    #
    cattr_accessor :dummy_date_for_time_type
    self.dummy_date_for_time_type = [ 2000, 1, 1 ]

    # Format tokens:
    #       y = year
    #       m = month
    #       d = day
    #       h = hour
    #       n = minute
    #       s = second
    #       u = micro-seconds
    #    ampm = meridian (am or pm) with or without dots (e.g. am, a.m, or a.m.)
    #       _ = optional space
    #      tz = Timezone abbreviation (e.g. UTC, GMT, PST, EST)
    #      zo = Timezone offset (e.g. +10:00, -08:00, +1000)
    #
    #   All other characters are considered literal. You can embed regexp in the
    #   format but no gurantees that it will remain intact. If you avoid the use
    #   of any token characters and regexp dots or backslashes as special characters
    #   in the regexp, it may well work as expected. For special characters use
    #   POSIX character clsses for safety.
    #
    # Repeating tokens:
    #       x = 1 or 2 digits for unit (e.g. 'h' means an hour can be '9' or '09')
    #      xx = 2 digits exactly for unit (e.g. 'hh' means an hour can only be '09')
    #
    # Special Cases:
    #      yy = 2 or 4 digit year
    #    yyyy = exactly 4 digit year
    #     mmm = month long name (e.g. 'Jul' or 'July')
    #     ddd = Day name of 3 to 9 letters (e.g. Wed or Wednesday)
    #       u = microseconds matches 1 to 6 digits
    #
    #   Any other invalid combination of repeating tokens will be swallowed up
    #   by the next lowest length valid repeating token (e.g. yyy will be
    #   replaced with yy)

    @@time_formats = [
      'hh:nn:ss',
      'hh-nn-ss',
      'h:nn',
      'h.nn',
      'h nn',
      'h-nn',
      'h:nn_ampm',
      'h.nn_ampm',
      'h nn_ampm',
      'h-nn_ampm',
      'h_ampm'
    ]

    @@date_formats = [
      'yyyy-mm-dd',
      'yyyy/mm/dd',
      'yyyy.mm.dd',
      'm/d/yy',
      'd/m/yy',
      'm\d\yy',
      'd\m\yy',
      'd-m-yy',
      'd.m.yy',
      'd mmm yy'
    ]

    @@datetime_formats = [
      'yyyy-mm-dd hh:nn:ss',
      'yyyy-mm-dd h:nn',
      'yyyy-mm-dd h:nn_ampm',
      'yyyy-mm-dd hh:nn:ss.u',
      'm/d/yy h:nn:ss',
      'm/d/yy h:nn_ampm',
      'm/d/yy h:nn',
      'd/m/yy hh:nn:ss',
      'd/m/yy h:nn_ampm',
      'd/m/yy h:nn',
      'ddd, dd mmm yyyy hh:nn:ss (zo|tz)', # RFC 822
      'ddd mmm d hh:nn:ss zo yyyy', # Ruby time string
      'yyyy-mm-ddThh:nn:ssZ', # iso 8601 without zone offset
      'yyyy-mm-ddThh:nn:sszo' # iso 8601 with zone offset
    ]


    # All tokens available for format construction. The token array is made of
    # token regexp, validation regexp and key for format proc mapping if any.
    # If the token needs no format proc arg then the validation regexp should
    # not have a capturing group, as all captured groups are passed to the
    # format proc.
    #
    # The token regexp should only use a capture group if 'look-behind' anchor
    # is required. The first capture group will be considered a literal and put
    # into the validation regexp string as-is. This is a hack.
    @@format_tokens = [
      { 'd'    => [ /(\A|[^d])d{1}(?=[^d])/, '(\d{1,2})', :day ] }, #/
      { 'ddd'  => [ /d{3,}/, '(\w{3,9})' ] },
      { 'dd'   => [ /d{2,}/, '(\d{2})',   :day ] },
      { 'mmm'  => [ /m{3,}/, '(\w{3,9})', :month ] },
      { 'mm'   => [ /m{2}/,  '(\d{2})',   :month ] },
      { 'm'    => [ /(\A|[^ap])m{1}/, '(\d{1,2})', :month ] },
      { 'yyyy' => [ /y{4,}/, '(\d{4})',   :year ] },
      { 'yy'   => [ /y{2,}/, '(\d{4}|\d{2})', :year ] },
      { 'hh'   => [ /h{2,}/, '(\d{2})',   :hour ] },
      { 'h'    => [ /h{1}/,  '(\d{1,2})', :hour ] },
      { 'nn'   => [ /n{2,}/, '(\d{2})',   :min ]  },
      { 'n'    => [ /n{1}/,  '(\d{1,2})', :min ] },
      { 'ss'   => [ /s{2,}/, '(\d{2})',   :sec ] },
      { 's'    => [ /s{1}/,  '(\d{1,2})', :sec ] },
      { 'u'    => [ /u{1,}/, '(\d{1,6})', :usec ] },
      { 'ampm' => [ /ampm/,  '((?:[aApP])\.?[mM]\.?)', :meridian ] },
      { 'zo'   => [ /zo/,    '([+-]\d{2}:?\d{2})', :offset ] },
      { 'tz'   => [ /tz/,    '(?:[A-Z]{1,4})' ] },
      { '_'    => [ /_/,     '\s?' ] }
    ]

    # Arguments which will be passed to the format proc if matched in the
    # time string. The key must be the key from the format tokens. The array
    # consists of the arry position of the arg, the arg name, and the code to
    # place in the time array slot. The position can be nil which means the arg
    # won't be placed in the array.
    #
    # The code can be used to manipulate the arg value if required, otherwise
    # should just be the arg name.
    #
    @@format_proc_args = {
      :year     => [0,   'y', 'unambiguous_year(y)'],
      :month    => [1,   'm', 'month_index(m)'],
      :day      => [2,   'd', 'd'],
      :hour     => [3,   'h', 'full_hour(h,md)'],
      :min      => [4,   'n', 'n'],
      :sec      => [5,   's', 's'],
      :usec     => [6,   'u', 'microseconds(u)'],
      :offset   => [7,   'z', 'offset_in_seconds(z)'],
      :meridian => [nil, 'md', nil]
    }

    class << self

      def compile_format_expressions
        @@time_expressions     = compile_formats(@@time_formats)
        @@date_expressions     = compile_formats(@@date_formats)
        @@datetime_expressions = compile_formats(@@datetime_formats)
      end

      # Loop through format expressions for type and call proc on matches. Allow
      # pre or post match strings to exist if strict is false. Otherwise wrap
      # regexp in start and end anchors.
      # Returns time array if matches a format, nil otherwise.
      def parse(string, type, options={})
        return string unless string.is_a?(String)
        options.reverse_merge!(:strict => true)

        sets = if options[:format]
          options[:strict] = true
          [ send("#{type}_expressions").assoc(options[:format]) ]
        else
          expression_set(type, string)
        end

        matches = nil
        processor = sets.each do |format, regexp, proc|
          full = /\A#{regexp}\Z/ if options[:strict]
          full ||= case type
          when :date     then /\A#{regexp}/
          when :time     then /#{regexp}\Z/
          when :datetime then /\A#{regexp}\Z/
          end
          break(proc) if matches = full.match(string.strip)
        end
        last = options[:include_offset] ? 8 : 7
        if matches
          values = processor.call(*matches[1..last])
          values[0..2] = dummy_date_for_time_type if type == :time
          return values
        end
      rescue
        nil
      end

      # Delete formats of specified type. Error raised if format not found.
      def remove_formats(type, *remove_formats)
        remove_formats.each do |format|
          unless self.send("#{type}_formats").delete(format)
            raise "Format #{format} not found in #{type} formats"
          end
        end
        compile_format_expressions
      end

      # Adds new formats. Must specify format type and can specify a :before
      # option to nominate which format the new formats should be inserted in
      # front on to take higher precedence.
      # Error is raised if format already exists or if :before format is not found.
      def add_formats(type, *add_formats)
        formats = self.send("#{type}_formats")
        options = {}
        options = add_formats.pop if add_formats.last.is_a?(Hash)
        before = options[:before]
        raise "Format for :before option #{format} was not found." if before && !formats.include?(before)

        add_formats.each do |format|
          raise "Format #{format} is already included in #{type} formats" if formats.include?(format)

          index = before ? formats.index(before) : -1
          formats.insert(index, format)
        end
        compile_format_expressions
      end

      # Removes formats where the 1 or 2 digit month comes first, to eliminate
      # formats which are ambiguous with the European style of day then month.
      # The mmm token is ignored as its not ambigous.
      def remove_us_formats
        us_format_regexp = /\Am{1,2}[^m]/
        date_formats.reject! { |format| us_format_regexp =~ format }
        datetime_formats.reject! { |format| us_format_regexp =~ format }
        compile_format_expressions
      end

      def full_hour(hour, meridian)
        hour = hour.to_i
        return hour if meridian.nil?
        if meridian.delete('.').downcase == 'am'
          raise if hour == 0 || hour > 12
          hour == 12 ? 0 : hour
        else
          hour == 12 ? hour : hour + 12
        end
      end

      def unambiguous_year(year)
        if year.length <= 2
          century = Time.now.year.to_s[0..1].to_i
          century -= 1 if year.to_i >= ambiguous_year_threshold
          year = "#{century}#{year.rjust(2,'0')}"
        end
        year.to_i
      end

      def month_index(month)
        return month.to_i if month.to_i.nonzero?
        abbr_month_names.index(month.capitalize) || month_names.index(month.capitalize)
      end

      def month_names
        defined?(I18n) ? I18n.t('date.month_names') : Date::MONTHNAMES
      end

      def abbr_month_names
        defined?(I18n) ? I18n.t('date.abbr_month_names') : Date::ABBR_MONTHNAMES
      end

      def microseconds(usec)
        (".#{usec}".to_f * 1_000_000).to_i
      end

      def offset_in_seconds(offset)
        sign = offset =~ /^-/ ? -1 : 1
        parts = offset.scan(/\d\d/).map {|p| p.to_f }
        parts[1] = parts[1].to_f / 60
        (parts[0] + parts[1]) * sign * 3600
      end

    private

      # Generate regular expression and processor from format string
      def generate_format_expression(string_format)
        regexp = string_format.dup
        order  = {}
        regexp.gsub!(/([\.\\])/, '\\\\\1') # escapes dots and backslashes

        format_tokens.each do |token|
          token_name = token.keys.first
          token_regexp, regexp_str, arg_key = *token.values.first

          # hack for lack of look-behinds. If has a capture group then is
          # considered an anchor to put straight back in the regexp string.
          regexp.gsub!(token_regexp) {|m| "#{$1}" + regexp_str }
          order[arg_key] = $~.begin(0) if $~ && !arg_key.nil?
        end

        return Regexp.new(regexp), format_proc(order)
      rescue
        raise "The following format regular expression failed to compile: #{regexp}\n from format #{string_format}."
      end

      # Generates a proc which when executed maps the regexp capture groups to a
      # proc argument based on order captured. A time array is built using the proc
      # argument in the position indicated by the first element of the proc arg
      # array.
      #
      def format_proc(order)
        arg_map = format_proc_args
        args = order.invert.sort.map {|p| arg_map[p[1]][1] }
        arr = [nil] * 7
        order.keys.each {|k| i = arg_map[k][0]; arr[i] = arg_map[k][2] unless i.nil? }
        proc_string = <<-EOL
          lambda {|#{args.join(',')}|
              md ||= nil
              [#{arr.map {|i| i.nil? ? 'nil' : i }.join(',')}].map {|i| i.is_a?(Float) ? i : i.to_i }
          }
        EOL
        eval proc_string
      end

      def compile_formats(formats)
        formats.map { |format| [ format, *generate_format_expression(format) ] }
      end

      # Pick expression set and combine date and datetimes for
      # datetime attributes to allow date string as datetime
      def expression_set(type, string)
        case type
        when :date
          date_expressions
        when :time
          time_expressions
        when :datetime
          # gives a speed-up for date string as datetime attributes
          if string.length < 11
            date_expressions + datetime_expressions
          else
            datetime_expressions + date_expressions
          end
        end
      end

    end
  end
end

ValidatesTimeliness::Formats.compile_format_expressions
